File: backend/app/services/document_service.py
import io
from pathlib import Path

import docx2txt
from fastapi import UploadFile
from langchain_text_splitters import RecursiveCharacterTextSplitter
from pypdf import PdfReader

from app.models.schemas import (
    DocumentAnalysisResponse,
    DocumentIngestResponse,
    KnowledgeDocumentRequest,
    KnowledgeIngestRequest,
    PdfTextExtractionResponse,
)
from app.i18n import t
from app.services.legal_service import legal_service
from app.settings import settings

SUPPORTED_EXTENSIONS = {".pdf", ".doc", ".docx", ".txt"}
MIN_EXTRACTED_CHARACTERS = 40


class DocumentService:
    async def extract_pdf_text(self, file: UploadFile) -> PdfTextExtractionResponse:
        return await self.extract_text(file)

    async def extract_text(self, file: UploadFile) -> PdfTextExtractionResponse:
        content = await file.read()
        filename = file.filename or "document"
        extension = Path(filename).suffix.lower()
        if extension not in SUPPORTED_EXTENSIONS:
            raise ValueError(t("document_unsupported"))
        if not content:
            raise ValueError(t("document_empty"))

        text = self._extract_text(filename, content)
        return PdfTextExtractionResponse(
            filename=filename,
            extractedCharacters=len(text),
            text=text,
        )

    async def analyze(self, file: UploadFile) -> DocumentAnalysisResponse:
        content = await file.read()
        filename = file.filename or t("uploaded_document")
        content_type = file.content_type or "application/octet-stream"
        issues = self._validate(filename, content)
        summary = t("document_review_warning")

        if not issues:
            try:
                text = self._extract_text(filename, content)
                summary = (
                    t("document_accepted", characters=len(text))
                    if text else t("document_no_readable_text")
                )
            except Exception as exc:
                issues.append(t("document_extraction_failed", detail=exc))
                summary = t("document_unreadable")

        return DocumentAnalysisResponse(
            filename=filename,
            size=len(content),
            contentType=content_type,
            detectedIssues=issues,
            summary=summary,
        )

    async def ingest(self, file: UploadFile, topic: str | None, court: str | None) -> DocumentIngestResponse:
        content = await file.read()
        filename = file.filename or t("uploaded_document")
        content_type = file.content_type or "application/octet-stream"
        warnings = self._validate(filename, content)
        if warnings:
            raise ValueError(" ".join(warnings))

        extracted_text = self._extract_text(filename, content)
        if len(extracted_text) < MIN_EXTRACTED_CHARACTERS:
            raise ValueError(t("document_insufficient_text"))

        cleaned_text = self._clean_text(extracted_text)
        splitter = RecursiveCharacterTextSplitter(chunk_size=1200, chunk_overlap=150)
        chunks = splitter.split_text(cleaned_text)
        document_topic = topic.strip() if topic and topic.strip() else Path(filename).stem
        documents = [
            KnowledgeDocumentRequest(
                sourceType="upload",
                court=court,
                topic=t("chunk_topic", topic=document_topic, index=index + 1, total=len(chunks)),
                summary=self._preview(chunk, 700),
                content=chunk,
            )
            for index, chunk in enumerate(chunks)
        ]
        ingest_response = legal_service.ingest_knowledge(KnowledgeIngestRequest(documents=documents))
        if ingest_response.indexed == 0:
            raise ValueError(ingest_response.message)

        return DocumentIngestResponse(
            filename=filename,
            size=len(content),
            contentType=content_type,
            extractedCharacters=len(cleaned_text),
            chunkCount=len(chunks),
            indexed=ingest_response.indexed,
            storage=ingest_response.storage,
            message=ingest_response.message,
            textPreview=self._preview(cleaned_text, 500),
            warnings=[],
        )

    def _validate(self, filename: str, content: bytes) -> list[str]:
        issues: list[str] = []
        extension = Path(filename).suffix.lower()
        max_bytes = settings.max_upload_mb * 1024 * 1024
        if not content:
            issues.append(t("document_empty"))
        if len(content) > max_bytes:
            issues.append(t("document_too_large", max_mb=settings.max_upload_mb))
        if extension not in SUPPORTED_EXTENSIONS:
            issues.append(t("document_unsupported"))
        return issues

    def _extract_text(self, filename: str, content: bytes) -> str:
        extension = Path(filename).suffix.lower()
        if extension == ".pdf":
            reader = PdfReader(io.BytesIO(content))
            return "\n".join(page.extract_text() or "" for page in reader.pages).strip()
        if extension == ".txt":
            return content.decode("utf-8", errors="ignore").strip()
        if extension in {".doc", ".docx"}:
            return docx2txt.process(io.BytesIO(content)).strip()
        return ""

    def _preview(self, content: str, length: int) -> str:
        return content if len(content) <= length else f"{content[:length]}..."

    def _clean_text(self, content: str) -> str:
        lines = [" ".join(line.split()) for line in content.splitlines()]
        paragraphs: list[str] = []
        current: list[str] = []
        for line in lines:
            if not line:
                if current:
                    paragraphs.append(" ".join(current))
                    current = []
                continue
            current.append(line)
        if current:
            paragraphs.append(" ".join(current))
        return "\n\n".join(paragraphs).strip()


document_service = DocumentService()

File: springboot-backend/legal-service/src/main/java/com/lawai/api/service/DocumentService.java
package com.lawai.api.service;

import com.lawai.api.dto.DocumentAnalysisResponse;
import com.lawai.api.dto.DocumentIngestResponse;
import com.lawai.api.dto.KnowledgeDocumentRequest;
import com.lawai.api.dto.KnowledgeIngestRequest;
import com.lawai.api.dto.KnowledgeIngestResponse;
import com.lawai.common.i18n.I18nMessages;
import org.apache.pdfbox.Loader;
import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.text.PDFTextStripper;
import org.apache.poi.hwpf.HWPFDocument;
import org.apache.poi.hwpf.extractor.WordExtractor;
import org.apache.poi.xwpf.usermodel.XWPFDocument;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;
import org.springframework.web.multipart.MultipartFile;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

@Service
public class DocumentService {

  private static final Set<String> SUPPORTED_EXTENSIONS = Set.of("pdf", "doc", "docx", "txt");
  private static final int MIN_EXTRACTED_CHARACTERS = 40;
  private static final int DEFAULT_CHUNK_SIZE = 1200;
  private static final int DEFAULT_CHUNK_OVERLAP = 150;

  private final int maxUploadMb;
  private final AiServiceClient aiServiceClient;
  private final I18nMessages i18n;

  public DocumentService(
      @Value("${app.max-upload-mb:25}") int maxUploadMb,
      AiServiceClient aiServiceClient,
      I18nMessages i18n
  ) {
    this.maxUploadMb = maxUploadMb;
    this.aiServiceClient = aiServiceClient;
    this.i18n = i18n;
  }

  public DocumentAnalysisResponse analyze(MultipartFile file) {
    String filename = safeFilename(file);
    String contentType = file.getContentType() == null ? "application/octet-stream" : file.getContentType();
    byte[] content = readBytes(file);
    List<String> issues = validate(filename, content);
    String summary = i18n.get("document.review-warning");

    if (issues.isEmpty()) {
      try {
        String text = extractText(filename, content);
        summary = text.isEmpty()
            ? i18n.get("document.no-readable-text")
            : i18n.get("document.accepted", text.length());
      } catch (Exception exc) {
        issues.add(i18n.get("document.extraction-failed", exc.getMessage()));
        summary = i18n.get("document.unreadable");
      }
    }

    return new DocumentAnalysisResponse(filename, content.length, contentType, issues, summary);
  }

  public DocumentIngestResponse ingest(MultipartFile file, String topic, String court) {
    String filename = safeFilename(file);
    String contentType = file.getContentType() == null ? "application/octet-stream" : file.getContentType();
    byte[] content = readBytes(file);
    List<String> warnings = validate(filename, content);
    if (!warnings.isEmpty()) {
      throw new IllegalArgumentException(String.join(" ", warnings));
    }

    String extractedText = extractText(filename, content);
    if (extractedText.length() < MIN_EXTRACTED_CHARACTERS) {
      throw new IllegalArgumentException(i18n.get("error.document-insufficient-text"));
    }

    List<String> chunks = splitIntoChunks(extractedText, DEFAULT_CHUNK_SIZE, DEFAULT_CHUNK_OVERLAP);
    String documentTopic = StringUtils.hasText(topic) ? topic.trim() : baseNameOf(filename);
    List<KnowledgeDocumentRequest> documents = new ArrayList<>();
    for (int index = 0; index < chunks.size(); index++) {
      String chunk = chunks.get(index);
      documents.add(new KnowledgeDocumentRequest(
          "upload",
          StringUtils.hasText(court) ? court : null,
          null,
          null,
          null,
          null,
          i18n.get("document.chunk-topic", documentTopic, index + 1, chunks.size()),
          preview(chunk, 700),
          chunk
      ));
    }

    KnowledgeIngestResponse ingestResponse = aiServiceClient.ingestKnowledge(new KnowledgeIngestRequest(documents));
    if (ingestResponse.indexed() == 0) {
      throw new IllegalArgumentException(ingestResponse.message());
    }

    return new DocumentIngestResponse(
        filename,
        content.length,
        contentType,
        extractedText.length(),
        chunks.size(),
        ingestResponse.indexed(),
        ingestResponse.storage(),
        ingestResponse.message(),
        preview(extractedText, 500),
        List.of()
    );
  }

  private List<String> validate(String filename, byte[] content) {
    List<String> issues = new ArrayList<>();
    String extension = extensionOf(filename);
    long maxBytes = (long) maxUploadMb * 1024L * 1024L;
    if (content.length == 0) {
      issues.add(i18n.get("error.document-empty"));
    }
    if (content.length > maxBytes) {
      issues.add(i18n.get("error.document-too-large", maxUploadMb));
    }
    if (!SUPPORTED_EXTENSIONS.contains(extension)) {
      issues.add(i18n.get("error.document-unsupported"));
    }
    return issues;
  }

  private String extractText(String filename, byte[] content) {
    String extension = extensionOf(filename);
    try {
      if ("pdf".equals(extension)) {
        try (PDDocument document = Loader.loadPDF(content)) {
          return new PDFTextStripper().getText(document).trim();
        }
      }
      if ("txt".equals(extension)) {
        return new String(content, StandardCharsets.UTF_8).trim();
      }
      if ("docx".equals(extension)) {
        try (XWPFDocument document = new XWPFDocument(new ByteArrayInputStream(content))) {
          return document.getParagraphs().stream()
              .map(paragraph -> paragraph.getText() == null ? "" : paragraph.getText())
              .reduce("", (left, right) -> left + "\n" + right)
              .trim();
        }
      }
      if ("doc".equals(extension)) {
        try (HWPFDocument document = new HWPFDocument(new ByteArrayInputStream(content));
             WordExtractor extractor = new WordExtractor(document)) {
          return extractor.getText().trim();
        }
      }
    } catch (IOException exc) {
      throw new IllegalStateException(i18n.get("error.document-read-failed", exc.getMessage()), exc);
    }
    return "";
  }

  private List<String> splitIntoChunks(String text, int chunkSize, int chunkOverlap) {
    List<String> chunks = new ArrayList<>();
    int index = 0;
    while (index < text.length()) {
      int end = Math.min(text.length(), index + chunkSize);
      chunks.add(text.substring(index, end));
      if (end >= text.length()) {
        break;
      }
      index = Math.max(end - chunkOverlap, index + 1);
    }
    return chunks;
  }

  private String preview(String content, int length) {
    return content.length() <= length ? content : content.substring(0, length) + "...";
  }

  private byte[] readBytes(MultipartFile file) {
    try {
      return file.getBytes();
    } catch (IOException exc) {
      throw new IllegalStateException(i18n.get("error.file-read-failed", exc.getMessage()), exc);
    }
  }

  private String safeFilename(MultipartFile file) {
    return StringUtils.hasText(file.getOriginalFilename()) ? file.getOriginalFilename() : i18n.get("document.uploaded-default");
  }

  private String extensionOf(String filename) {
    int index = filename.lastIndexOf('.');
    return index < 0 ? "" : filename.substring(index + 1).toLowerCase();
  }

  private String baseNameOf(String filename) {
    int separator = Math.max(filename.lastIndexOf('/'), filename.lastIndexOf('\\'));
    String name = separator < 0 ? filename : filename.substring(separator + 1);
    int dot = name.lastIndexOf('.');
    return dot < 0 ? name : name.substring(0, dot);
  }
}
